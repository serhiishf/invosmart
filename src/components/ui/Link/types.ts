import { AnchorHTMLAttributes } from 'react';

export interface LinkProps extends AnchorHTMLAttributes<HTMLAnchorElement> {
  label: string;
  href: string;
  /**
   * Indicates whether the link is internal. If false and target="_blank" is used,
   * add `rel="noopener noreferrer"` to prevent performance and security issues.
   */
  isInternalLink?: boolean;
}
