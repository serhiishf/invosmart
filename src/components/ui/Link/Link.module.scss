.link {
  color: $linkColor;
  position: relative;
  width: fit-content;

  &:visited {
    color: $linkColor;
  }
  &:hover {
    text-decoration: underline;
  }
}
