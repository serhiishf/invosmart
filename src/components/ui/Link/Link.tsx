import styles from './Link.module.scss';
import { LinkProps } from './types';

function Link(props: LinkProps) {
  const { children, href, ...rest } = props;
  return (
    <a href={href} className={styles.link} {...rest}>
      {children}
    </a>
  );
}

export default Link;
