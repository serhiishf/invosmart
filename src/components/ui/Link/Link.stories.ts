import type { Meta, StoryObj } from '@storybook/react';
import Link from './Link';

const meta = {
  title: 'components/UI/Link',
  component: Link,
  tags: ['autodocs'],
} satisfies Meta<typeof Link>;

export default meta;

type Story = StoryObj<typeof meta>;
export const Default: Story = {
  args: {
    label: 'Example link',
    href: 'https://example.com',
    target: '_blank',
  },
};
