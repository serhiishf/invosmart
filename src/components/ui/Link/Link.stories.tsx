import type { Meta, StoryObj } from '@storybook/react';
import Link from './Link';

const meta = {
  title: 'components/UI/Link',
  component: Link,
  tags: ['autodocs'],
} satisfies Meta<typeof Link>;

export default meta;

type Story = StoryObj<typeof meta>;
export const Default: Story = {
  args: {
    label: 'Example link',
    href: 'https://example.com',
    target: '_blank',
  },
};

export const PseudoStates: Story = {
  args: {
    label: 'Example link',
    href: 'https://example.com',
    target: '_blank',
  },

  render: (args) => (
    <div style={{ display: 'flex', flexDirection: 'column', gap: '20px' }}>
      <Link {...args} id="one" label="Default link" href="https://some-random-address.com" />
      <Link {...args} id="two" label="Visited link" />
      <Link {...args} id="three" label="Hovered link" />
    </div>
  ),

  parameters: {
    pseudo: {
      unvisited: '#one',
      visited: '#two',
      hover: '#three',
    },
    controls: {
      disable: true, // This disables all controls for this story
    },
  },
};
