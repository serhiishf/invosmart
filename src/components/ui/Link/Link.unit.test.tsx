import { render, screen } from '@testing-library/react';
import Link from './Link';

describe('Link', () => {
  it('should render correctly', () => {
    render(<Link label="Some link" href="http://example.com" />);
    const link = screen.getByRole('link', { name: 'Some link' });
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', 'http://example.com');
  });

  it('should accept the attributes of the html <a> link include attribute "target"', () => {
    render(<Link label="Target _blank link" href="http://example.com" target="_blank" />);
    const link = screen.getByRole('link', { name: 'Target _blank link' });
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('target', '_blank');
  });

  it('should add rel="noopener noreferrer" when target is "_blank" and not an internal link', () => {
    render(<Link label="Not internal link" href="http://example.com" target="_blank" />);
    const link = screen.getByRole('link', { name: 'Not internal link' });
    expect(link).toHaveAttribute('rel', 'noopener noreferrer');
  });

  it('should matches the snapshot with label="Example link" and href="http://example.com"', () => {
    const { asFragment } = render(<Link label="Example link" href="http://example.com" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should mathces the snapshot with classname passed in props', () => {
    const { asFragment } = render(
      <Link label="Example link" href="http://example.com" className="additionalClassName" />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
