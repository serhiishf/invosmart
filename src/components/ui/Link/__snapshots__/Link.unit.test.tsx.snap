// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link > should matches the snapshot with label="Example link" and href="http://example.com" 1`] = `
<DocumentFragment>
  <a
    class="_link_1f0167"
    href="http://example.com"
  >
    Example link
  </a>
</DocumentFragment>
`;

exports[`Link > should mathces the snapshot with classname passed in props 1`] = `
<DocumentFragment>
  <a
    class="_link_1f0167 additionalClassName"
    href="http://example.com"
  >
    Example link
  </a>
</DocumentFragment>
`;
